### PR TITLE
Locus Specific Bloom Features

### PR DESCRIPTION
**Pull request by DLu**
(plus one thing that just annoys me)

 * Opens ~pull~ merge requests for gitlab! 
 * Will automatically attempt to use the `http://gitlab-ci-token:TOKEN@gitlab.locusbots.io/locusrobotics/REPOSITORY-release.git` as the release repository if it exists.
 * If not specified, assumes that `track` is the same as `rosdistro`

The gitlab portions will only work if you have `pip install python-gitlab`. The github portions use a custom interface to the api (i.e. not pygithub or similar). I don't know if that's to avoid the dependency. Furthermore, the gitlab portions are probably not ready to be merged to upstream since additional logic is probably needed for seeking out the right token, rather than assuming it is installed in `~/.git-tokens`

### CODE REVIEW

File: bloom/commands/release.py [hunk 1]
```diff
@@ -707,6 +749,24 @@ def get_github_interface(quiet=False):
     return gh
 
 
+def get_gitlab_interface(server, quiet=False):
+    global _gl
+    if _gl is not None:
+        return _gl
+    # Make sure we can import gitlab
+    try:
+        import gitlab
```
Comment (paulbovbel) [inline bloom/commands/release.py]: I think you can safely import this up top if you add it to https://github.com/locusrobotics/bloom/blob/locus-master/setup.py#L6, that should ensure dev machines pull the library down as well.

**Comment by DLu** [comment]
P.S. I've made two successful PRs with this code. 
e.g. http://gitlab.locusbots.io/locusrobotics/rosdistro/merge_requests/10

**Review by paulbovbel — CHANGES_REQUESTED** [review]
LGTM, pending the requirement change.

If you wanted to take this upstream and reduce the fork factor, you could remove the locus-convenience stuff (maybe have some kind of opt-in configuration options for repository url templating), and detect a gitlab rosdistro (maybe `if 'gitlab' in server`).

**Review by ayrton04 — APPROVED** [review]
Wow! That's really handy. So now it behaves as it does when you bloom from a GitHub-based rosdistro?

**Review by paulbovbel — APPROVED** [review]

**Comment by DLu** [comment]
@ayrton04 Pretty much, modulo where the credentials are saved and how the API is accessed. 

@ayrton04 @paulbovbel 
What if there was an environment variable that was `BLOOM_RELEASE_REPO_BASE` that we could set to `http://gitlab-ci-XXXX@gitlab.locusbots.io/locusrobotics/` and I personally might set to `https://github.com/wu-robotics/` and then `$REPOSITORY-release.git` would be appended onto the end. Does that sound reasonable?

Then we could store the gitlab access key in `.config/bloom` like the github access stuff is, and remove the special git-tokens variable. 

I should note that the one other thing I did in the `locus-bloom` script was to make sure the github token was prepended onto the URLs that go into rosdistro. Given your familiarity with the build farm, how important do you consider that?

**Comment by paulbovbel** [comment]
That all sounds very reasonable, especially if just sets the default and can be overriden during the `--edit` phase. You may want to consider just dumping more keys into `~/.config/bloom` rather than dealing with environments.

If this is just for us, then ++1, if it's for upstream, you may just want to make a PR and work through it with '@'nuclearsandwich - I'm sure upstream would love to enable gitlab MRs, but they have a lot more use cases to consider than we do.

**Comment by DLu** [comment]
The last two commits remove the git-tokens Locus-specific logic, and updates the gitlab interaction to be more in the same style of the github interaction. Also removes our specific patterns in favor of environment vars. Please take another look.

File: bloom/commands/release.py [hunk 2]
```diff
@@ -680,20 +680,28 @@ def get_gl_info(url):
...
         with open(oauth_config_path, 'r') as f:
             config = json.loads(f.read())
-            token = config.get('oauth_token', None)
-            username = config.get('github_user', None)
-            if token and username:
-                return Github(username, auth=auth_header_from_oauth_token(token), token=token)
+    return config, oauth_config_path
+
+
+
```
Comment (ayrton04) [inline bloom/commands/release.py]: Is this one too many line breaks?

**Review by ayrton04 — APPROVED** [review]

**Review by paulbovbel — APPROVED** [review]

File: bloom/commands/release.py (hunk 2 above)
Comment (DLu) [reply bloom/commands/release.py]: ERROR: Too many linting errors for me to notice this one.

Thanks.

**Comment by DLu** [comment]
Thank you. PR opened upstream https://github.com/ros-infrastructure/bloom/pull/458

**Pull request merged**